File: 0045-jump-game-ii/0045-jump-game-ii.cpp
class Solution {
public:
    int dfs(int i, vector<int> &nums, vector<int> &dp){
        if(i >= nums.size() - 1) return 0;

        if(dp[i] != -1) return dp[i];
        int mini = 1e9;
        for(int j = i + 1; j <= i + nums[i]; j++){
            mini = min(mini, 1 + dfs(j,nums,dp));
        }   
        return dp[i] = mini;
    }
    int jump(vector<int>& nums) {
        vector<int> dp(nums.size(), -1);
        return dfs(0,nums,dp);
    }
};

File: 0045-jump-game-ii/0045-jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        # maintain a range by two variables 
        l = far = jumps = r = 0
        while r < len(nums) - 1:
            far = 0
            for i in range(l,r+1):
                far = max(far, i + nums[i])
            
            l = r + 1
            r = far
            jumps += 1
        return jumps